--- star-1.5/DEFAULTS/Defaults.linux.newMake	2003-02-16 01:01:48.000000000 +0100
+++ star-1.5/DEFAULTS/Defaults.linux	2004-11-22 12:29:02.973507632 +0100
@@ -8,8 +8,8 @@
 # Compiler stuff
 #
 ###########################################################################
-DEFCCOM=	cc
-#DEFCCOM=	gcc
+#DEFCCOM=	cc
+DEFCCOM=	gcc
 ###########################################################################
 #
 # If the next line is commented out, compilation is done with max warn level
@@ -18,21 +18,21 @@
 ###########################################################################
 CWARNOPTS=
 
-DEFINCDIRS=	$(SRCROOT)/include /usr/src/linux/include
-LDPATH=		-L/opt/schily/lib
-RUNPATH=	-R $(INS_BASE)/lib -R /opt/schily/lib -R $(OLIBSDIR)
+DEFINCDIRS=	$(SRCROOT)/include
+LDPATH=		
+RUNPATH=	
 
 ###########################################################################
 #
 # Installation config stuff
 #
 ###########################################################################
-INS_BASE=	/opt/schily
-INS_KBASE=	/
+INS_BASE=	$(RPM_INSTALLDIR)/usr/
+INS_KBASE=	$(RPM_INSTALLDIR)/
 #
 DEFUMASK=	002
 #
 DEFINSMODEF=	444
 DEFINSMODEX=	755
-DEFINSUSR=	bin
-DEFINSGRP=	bin
+DEFINSUSR=	root
+DEFINSGRP=	root
